### PR TITLE
MAINT: refactor unravel_index for code repetition

### PR DESCRIPTION
**Pull request by jaimefrio**
Adding a little bit of extra accounting in the inner loop allows joining the two specialized branches into a single one. The performance loss, if any, is within the measurement error.

### CODE REVIEW

**Comment by seberg** [comment]
Sorry to bug you with it, but since you actually touched that function.... It at least used to not be able to correctly throw an error on unravelling an index into an empty shapes.

**Comment by jaimefrio** [comment]
I just changed the comment before `ravel_multi_index_loop`, which was erroneously saying the function was the inner loop for `unravel_index`. All the other changes are `unravel_index` related.

File: numpy/core/src/multiarray/compiled_base.c [hunk 1]
```diff
@@ -999,84 +999,57 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
+unravel_index_loop(int unravel_ndim, npy_intp *unravel_dims,
+                   npy_intp unravel_size, npy_intp count,
+                   char *indices, npy_intp indices_stride,
+                   npy_intp *coords, NPY_ORDER order)
 {
-    int i;
-    char invalid;
-    npy_intp val;
+    int i, idx;
+    int idx_start = (order == NPY_CORDER) ? unravel_ndim - 1: 0;
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Should have a note that only `NPY_CORDER` and `NPY_FORTRANORDER` are valid values for `order`.

File: numpy/core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -999,84 +999,57 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
+    int idx_step = (order == NPY_CORDER) ? -1 : 1;
+    char invalid = 0;
 
     NPY_BEGIN_ALLOW_THREADS;
-    invalid = 0;
+    /* NPY_KEEPORDER or NPY_ANYORDER have no meaning in this setting */
+    assert (order == NPY_CORDER || order == NPY_FORTRANORDER);
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: OK, good to have this check here.

File: numpy/core/src/multiarray/compiled_base.c [hunk 3]
```diff
@@ -999,84 +999,57 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
     while (count--) {
-        val = *(npy_intp *)indices;
+        npy_intp val = *(npy_intp *)indices;
         if (val < 0 || val >= unravel_size) {
             invalid = 1;
             break;
         }
+        idx = idx_start;
         for (i = 0; i < unravel_ndim; ++i) {
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: A common construct for repeats is
```
for (i = unravel_ndim; i != 0; --i)
```

File: numpy/core/src/multiarray/compiled_base.c [hunk 4]
```diff
@@ -999,84 +999,57 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
-            *coords++ = val % unravel_dims[i];
-            val /= unravel_dims[i];
+            coords[idx] = val % unravel_dims[idx];
+            val /= unravel_dims[idx];
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Might want to reverse the order here as the remainder may be produced as part of the division. See http://stackoverflow.com/questions/7070346/c-best-way-to-get-integer-division-and-remainder.

Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Might need to assign `unravel_dims[idx]` to a constant variable. Depends on how smart the compiler is.

File: numpy/core/src/multiarray/compiled_base.c [hunk 5]
```diff
@@ -999,84 +999,57 @@ arr_ravel_multi_index(PyObject *self, PyObject *args, PyObject *kwds)
...
+            idx += idx_step;
         }
+        coords += unravel_ndim;
         indices += indices_stride;
     }
     NPY_END_ALLOW_THREADS;
     if (invalid) {
         PyErr_SetString(PyExc_ValueError,
-              "invalid entry in index array");
+                    "invalid entry in index array");
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Heh, and I usually outdent to the original. We don't have a style exception for this, but the errors are almost always style exceptions.

File: numpy/core/src/multiarray/compiled_base.c [hunk 6]
```diff
@@ -1161,64 +1134,35 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
...
-            npy_intp *strides;
-            npy_intp *countptr, count;
-            npy_intp *coordsptr = (npy_intp *)PyArray_DATA(ret_arr);
+        dataptr = NpyIter_GetDataPtrArray(iter);
+        strides = NpyIter_GetInnerStrideArray(iter);
+        countptr = NpyIter_GetInnerLoopSizePtr(iter);
 
-            iternext = NpyIter_GetIterNext(iter, NULL);
-            if (iternext == NULL) {
+        do {
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Don't see many of these ;)

File: numpy/core/src/multiarray/compiled_base.c [hunk 7]
```diff
@@ -1161,64 +1134,35 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
...
-                coordsptr += count*dimensions.len;
-            } while(iternext(iter));
-        }
-    }
-    else {
-        PyErr_SetString(PyExc_ValueError,
-                        "only 'C' or 'F' order is permitted");
-        goto fail;
+            coordsptr += count * dimensions.len;
+        } while(iternext(iter));
```
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Probably want a space after `while`.

File: numpy/core/src/multiarray/compiled_base.c (hunk 1 above)
Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: As part of function documentation.

Comment (charris) [inline numpy/core/src/multiarray/compiled_base.c]: Wonder if it is worth doing Doxygen style function documentation here? We really should be taking steps in that direction.

**Comment by charris** [comment]
LGTM modulo some style nitpicks.

**Comment by eric-wieser** [comment]
Looks like some good cleanup, but needs a rebase.

**Comment by mattip** [comment]
@jaimefrio may I push the rebase/cleanups to your branch?

**Comment by mattip** [comment]
I rebased this off master and opened #13446 to hopefully get this cleanup merged soon. I will tentatively close this one, but if @jaimefrio wishes to continue, feel free to close the other and reopen this

**Pull request closed without merging**